README.md
# Nepali Calendar 

Nepali Calendar with both English and Devanagari language and with custom styles and colors. Extension of [nepali_date_picker](https://pub.dev/packages/nepali_date_picker)

### Upcoming Features
Nepali to English Calendar Conversion + English Calendar View

![Loading Demo](demo.gif)

### Salient Features
* Nepali Calendar
* Events
* Nepali Calendar in both English and Devanagiri Text
* Supports from 1970 B.S. to 2100 B.S.
* English Calendar Conversion [Upcoming]

<img src="ss/ss1.png" width="400px"> <img src="ss/ss2.png" width="400px">

## Installation

Add package to pubspec.yaml.

```dart
nepali_english_calendar: ^1.0.1+1
```
## Usages

Import the package.

```dart
import 'package:nepali_english_calendar/nepali_english_calendar.dart'
```
Add NepaliCalendar to the widget tree

```dart
    NepaliCalendar(
        monthYearPickerStyle:
            TextStyle(color: Colors.red, fontWeight: FontWeight.bold, fontSize: 24),
        //Color to left right button
        rightLeftButtonColor: Colors.red,
        //Styles to Week Row
        weekHeaderStyle:
            TextStyle(color: Colors.blue, fontWeight: FontWeight.bold, fontSize: 26),
        initialDate: NepaliDateTime.now(),
        firstDate: NepaliDateTime(2070),
        lastDate: NepaliDateTime(2090),
        onDateChanged: (date) => _selectedDate.value = date,
        dayBuilder: (dayToBuild) {
        return Stack(
            children: <Widget>[
            Center(
                child: Text(
                NepaliUtils().language == Language.english
                    ? '${dayToBuild.day}'
                    : NepaliUnicode.convert('${dayToBuild.day}'),
                style: Theme.of(context)
                    .textTheme
                    .bodyMedium!
                    .copyWith(fontWeight: FontWeight.bold),
                ),
            ),
            if (events.any((event) => _dayEquals(event.date, dayToBuild)))
                Align(
                alignment: Alignment.bottomCenter,
                child: Container(
                    width: 6,
                    height: 6,
                    decoration:
                        BoxDecoration(shape: BoxShape.circle, color: Colors.red),
                ),
                )
            ],
        );
        },
        selectedDayDecoration: BoxDecoration(
        color: Colors.deepOrange,
        shape: BoxShape.circle,
        ),
        todayDecoration: BoxDecoration(
        gradient: LinearGradient(colors: [Colors.yellow, Colors.orange]),
        shape: BoxShape.circle,
        ),
    ),
```

## License

```
MIT License

Copyright (c) [2023] [Lakshydeep Vikram Sah]

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
```

## Support and feedback

If I miss to add some feature, feel free to open a ticket or contribute!
You can reach me here:
[LinkedIn](https://www.linkedin.com/in/lakshydeep-14/) | 
[GitHub](https://github.com/lakshydeep-14) | 
[Medium](https://lakshydeep-14.medium.com/) | 
[Google DevLibrary](https://devlibrary.withgoogle.com/authors/lakshydeep-14)


CHANGELOG.md
## 1.0.1+1
* docs updated.

## 1.0.1
* docs updated.

## 1.0.0
* initial release.


lib/src/date_utils.dart
// Copyright 2023 Lakshydeep Vikram Sah. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:flutter/material.dart';
import 'package:nepali_utils/nepali_utils.dart';

import 'date_picker_common.dart';

/// Returns a [NepaliDateTime] with just the date of the original, but no time set.
NepaliDateTime dateOnly(NepaliDateTime date) {
  return NepaliDateTime(date.year, date.month, date.day);
}

/// Returns true if the two [NepaliDateTime] objects have the same day, month, and
/// year, or are both null.
bool isSameDay(NepaliDateTime? dateA, NepaliDateTime? dateB) {
  return dateA?.year == dateB?.year &&
      dateA?.month == dateB?.month &&
      dateA?.day == dateB?.day;
}

/// Returns true if the two [NepaliDateTime] objects have the same month, and
/// year, or are both null.
bool isSameMonth(NepaliDateTime? dateA, NepaliDateTime? dateB) {
  return dateA?.year == dateB?.year && dateA?.month == dateB?.month;
}

/// Determines the number of months between two [NepaliDateTime] objects.
int monthDelta(NepaliDateTime startDate, NepaliDateTime endDate) {
  return (endDate.year - startDate.year) * 12 + endDate.month - startDate.month;
}

/// Returns a [NepaliDateTime] with the added number of months and truncates any day
/// and time information.
NepaliDateTime addMonthsToMonthDate(NepaliDateTime monthDate, int monthsToAdd) {
  var _year = monthDate.year;
  var _month = monthDate.month + monthsToAdd;

  _year += (_month - 1) ~/ 12;
  _month = _month % 12;
  if (_month == 0) _month = 12;
  return NepaliDateTime(_year, _month);
}

/// Computes the offset from the first day of the week that the first day of
/// the [month] falls on.
int firstDayOffset(int year, int month) {
  return NepaliDateTime(year, month).weekday - 1;
}

/// Returns the number of days in a month.
int getDaysInMonth(int year, int month) {
  return NepaliDateTime(year, month).totalDays;
}

/// Returns a locale-appropriate string to describe the start of a date range.
///
/// If `startDate` is null, then it defaults to 'Start Date', otherwise if it
/// is in the same year as the `endDate` then it will use the short month
/// day format (i.e. 'Asr 21'). Otherwise it will return the short date format
/// (i.e. 'Asr 21, 2077').
String formatRangeStartDate(MaterialLocalizations localizations,
    NepaliDateTime? startDate, NepaliDateTime? endDate) {
  return startDate == null
      ? localizations.dateRangeStartLabel
      : (endDate == null || startDate.year == endDate.year)
          ? NepaliDateFormat('MMMM d').format(startDate)
          : NepaliDateFormat.yMd().format(startDate);
}

/// Returns an locale-appropriate string to describe the end of a date range.
///
/// If `endDate` is null, then it defaults to 'End Date', otherwise if it
/// is in the same year as the `startDate` and the `currentDate` then it will
/// just use the short month day format (i.e. 'Asr 21'), otherwise it will
/// include the year (i.e. 'Asr 21, 2077').
String formatRangeEndDate(
    MaterialLocalizations localizations,
    NepaliDateTime? startDate,
    NepaliDateTime? endDate,
    NepaliDateTime currentDate) {
  return endDate == null
      ? localizations.dateRangeEndLabel
      : (startDate != null &&
              startDate.year == endDate.year &&
              startDate.year == currentDate.year)
          ? NepaliDateFormat('MMMM d').format(endDate)
          : NepaliDateFormat.yMd().format(endDate);
}

/// Returns a [NepaliDateTimeRange] with the dates of the original without any times set.
NepaliDateTimeRange datesOnly(NepaliDateTimeRange range) {
  return NepaliDateTimeRange(
      start: dateOnly(range.start), end: dateOnly(range.end));
}


lib/src/month_formattor.dart
// Copyright 2023 Lakshydeep Vikram Sah. All rights reserved.

import 'package:nepali_utils/nepali_utils.dart';

/// Formats month to literal form
String formattedMonth(
  int month, [
  Language? language,
]) {
  final _isEnglish = (language ?? NepaliUtils().language) == Language.english;
  switch (month) {
    case 1:
      return _isEnglish ? 'Baisakh' : 'बैशाख';
    case 2:
      return _isEnglish ? 'Jestha' : 'जेष्ठ';
    case 3:
      return _isEnglish ? 'Ashad' : 'आषाढ';
    case 4:
      return _isEnglish ? 'Shrawan' : 'श्रावण';
    case 5:
      return _isEnglish ? 'Bhadra' : 'भाद्र';
    case 6:
      return _isEnglish ? 'Ashwin' : 'आश्विन';
    case 7:
      return _isEnglish ? 'Kartik' : 'कार्त्तिक';
    case 8:
      return _isEnglish ? 'Mangsir' : 'मङ्सिर';
    case 9:
      return _isEnglish ? 'Poush' : 'पौष';
    case 10:
      return _isEnglish ? 'Magh' : 'माघ';
    case 11:
      return _isEnglish ? 'Falgun' : 'फाल्गुण';
    case 12:
      return _isEnglish ? 'Chaitra' : 'चैत्र';
    default:
      return '';
  }
}


lib/src/date_picker_common.dart
// Copyright 2023 Lakshydeep Vikram Sah. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:flutter/foundation.dart';
import 'package:nepali_utils/nepali_utils.dart';

/// Signature for predicating dates for enabled date selections.
typedef SelectableDayPredicate = bool Function(NepaliDateTime day);

/// Encapsulates a start and end [NepaliDateTime] that represent the range of dates
/// between them.
@immutable
class NepaliDateTimeRange {
  /// Creates a date range for the given start and end [NepaliDateTime].
  ///
  /// [start] and [end] must be non-null.
  const NepaliDateTimeRange({
    required this.start,
    required this.end,
  });

  /// The start of the range of dates.
  final NepaliDateTime start;

  /// The end of the range of dates.
  final NepaliDateTime end;

  /// Returns a [Duration] of the time between [start] and [end].
  ///
  /// See [DateTime.difference] for more details.
  Duration get duration => end.difference(start);

  @override
  bool operator ==(Object other) {
    if (other.runtimeType != runtimeType) return false;
    return other is NepaliDateTimeRange &&
        other.start == start &&
        other.end == end;
  }

  @override
  int get hashCode => Object.hash(start, end);

  @override
  String toString() => '$start - $end';
}


example/lib/main.dart
// Copyright 2023 Lakshydeep Vikram Sah. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:flutter/material.dart';
import 'package:nepali_utils/nepali_utils.dart';
import 'package:nepali_english_calendar/nepali_english_calendar.dart';

Color red = const Color(0xffDC143C);
Color blue = const Color(0xff003893);
void main() => runApp(const MyApp());

/// MyApp
class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    NepaliUtils().language == Language.nepali;
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        appBar: AppBar(
          backgroundColor: blue,
          title: const Text("🇳🇵Nepali Calendar🇳🇵"),
          centerTitle: true,
          actions: [
            IconButton(
              icon: Text(
                NepaliUtils().language == Language.english ? 'ने' : 'En',
                style: Theme.of(context)
                    .textTheme
                    .titleLarge
                    ?.copyWith(color: Colors.white),
              ),
              onPressed: () {
                NepaliUtils().language =
                    NepaliUtils().language == Language.english
                        ? Language.nepali
                        : Language.english;
                setState(() {});
              },
            ),
          ],
        ),
        body: HomePage(
          nepaliUtils: NepaliUtils(NepaliUtils().language == Language.english
              ? Language.english
              : Language.nepali),
        ),
      ),
    );
  }
}

class HomePage extends StatefulWidget {
  final NepaliUtils nepaliUtils;

  const HomePage({super.key, required this.nepaliUtils});

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  final ValueNotifier<NepaliDateTime> _selectedDate =
      ValueNotifier(NepaliDateTime.now());

  /// Events
  final List<Event> events = [
    Event(date: NepaliDateTime.now(), eventTitles: ['Today 1', 'Today 2']),
    Event(
        date: NepaliDateTime.now().add(const Duration(days: 30)),
        eventTitles: ['Holiday 1', 'Holiday 2']),
    Event(
        date: NepaliDateTime.now().subtract(const Duration(days: 5)),
        eventTitles: ['Event 1', 'Event 2']),
    Event(
        date: NepaliDateTime.now().add(const Duration(days: 8)),
        eventTitles: ['Seminar 1', 'Seminar 2']),
  ];

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        NepaliCalendar(
          language: widget.nepaliUtils, //by default Nepali
          monthYearPickerStyle:
              TextStyle(color: red, fontWeight: FontWeight.bold, fontSize: 24),
          //Color to left right button
          rightLeftButtonColor: Colors.red,
          //Styles to Week Row
          weekHeaderStyle:
              TextStyle(color: blue, fontWeight: FontWeight.bold, fontSize: 26),
          initialDate: NepaliDateTime.now(),
          firstDate: NepaliDateTime(2070),
          lastDate: NepaliDateTime(2090),
          onDateChanged: (date) => _selectedDate.value = date,
          dayBuilder: (dayToBuild) {
            return Stack(
              children: <Widget>[
                Center(
                  child: Text(
                    NepaliUtils().language == Language.english
                        ? '${dayToBuild.day}'
                        : NepaliUnicode.convert('${dayToBuild.day}'),
                    style: Theme.of(context)
                        .textTheme
                        .titleMedium!
                        .copyWith(fontWeight: FontWeight.bold),
                  ),
                ),
                if (events.any((event) => _dayEquals(event.date, dayToBuild)))
                  Align(
                    alignment: Alignment.bottomCenter,
                    child: Container(
                      width: 6,
                      height: 6,
                      decoration:
                          BoxDecoration(shape: BoxShape.circle, color: red),
                    ),
                  )
              ],
            );
          },
          selectedDayDecoration: const BoxDecoration(
            color: Colors.deepOrange,
            shape: BoxShape.circle,
          ),
          todayDecoration: const BoxDecoration(
            gradient: LinearGradient(colors: [Colors.yellow, Colors.orange]),
            shape: BoxShape.circle,
          ),
        ),
        Expanded(
          child: ValueListenableBuilder<NepaliDateTime>(
            valueListenable: _selectedDate,
            builder: (context, date, _) {
              Event? event;
              try {
                event = events.firstWhere((e) => _dayEquals(e.date, date));
              } on StateError {
                event = null;
              }

              if (event == null) {
                return const Center(
                  child: Text('No Events'),
                );
              }

              return ListView.separated(
                itemCount: event.eventTitles.length,
                itemBuilder: (context, index) => ListTile(
                  leading: TodayWidget(
                    today: date,
                  ),
                  title: Text(event!.eventTitles[index]),
                  onTap: () {},
                ),
                separatorBuilder: (context, _) => const Divider(),
              );
            },
          ),
        ),
      ],
    );
  }

  bool _dayEquals(NepaliDateTime? a, NepaliDateTime? b) =>
      a != null &&
      b != null &&
      a.toIso8601String().substring(0, 10) ==
          b.toIso8601String().substring(0, 10);
}

///
class TodayWidget extends StatelessWidget {
  ///
  final NepaliDateTime today;

  ///
  const TodayWidget({
    Key? key,
    required this.today,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Material(
      elevation: 4,
      shape: RoundedRectangleBorder(
        borderRadius: BorderRadius.circular(10),
      ),
      child: SizedBox(
        width: 60,
        height: 60,
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            DecoratedBox(
              decoration: BoxDecoration(
                color: red,
                borderRadius: const BorderRadius.only(
                  topLeft: Radius.circular(10),
                  topRight: Radius.circular(10),
                ),
              ),
              child: Padding(
                padding: const EdgeInsets.symmetric(vertical: 2),
                child: Text(
                  NepaliDateFormat.EEEE()
                      .format(today)
                      .substring(0, 3)
                      .toUpperCase(),
                  style: Theme.of(context)
                      .textTheme
                      .bodyLarge
                      ?.copyWith(color: Colors.white),
                  textAlign: TextAlign.center,
                ),
              ),
            ),
            const Spacer(),
            Text(
              NepaliDateFormat.d().format(today),
              style: Theme.of(context).textTheme.titleLarge,
              textAlign: TextAlign.center,
            ),
            const Spacer(),
          ],
        ),
      ),
    );
  }
}

///
class Event {
  ///
  final NepaliDateTime date;

  ///
  final List<String> eventTitles;

  ///
  Event({required this.date, required this.eventTitles});
}
